device-connectors/src/testflinger_device_connectors/devices/zapper_iot/__init__.py

```diff
@@ -56,16 +60,44 @@ def _validate_configuration(
         if provision_plan:
 
             try:
-                validate_provision_plan(provision_plan)
-
                 # Make sure the user created at provision time is
                 # the same used during the test phase.
                 provision_plan["config"]["username"] = username
                 provision_plan["config"]["password"] = password
 
+                # If the initial login is using console-conf, we need to
+                # check if the user has provided an ubuntu_sso_email.
+                # If not, we raise an error.
+                # ubuntu_sso_email will be used to set the username for the
+                # initial login.
+                run_stages = provision_plan["run_stage"]
+                for stage in run_stages:
+                    if (
+                        isinstance(stage, dict)
+                        and "initial_login" in stage.keys()
+                        and stage["initial_login"].get("method")
+                        == "console-conf"
+                    ):
+                        if ubuntu_sso_email:
+                            provision_plan["config"][
+                                "username"
+                            ] = ubuntu_sso_email
+                            break
+                        else:
+                            raise ValueError(
+                                "ubuntu_sso_email is required "
+                                "when initial login using console-conf"
+                            )
```

Comment:
In cases like this, I personally enjoy using `contextlib.suppress` which will let us iterate w/o bothering with exceptions until we find exactly what we need.
```suggestion
                    with contextlib.suppress(KeyError):
                        if stage["initial_config"]["method"] == "console-conf":
                            if not ubuntu_sso_email:
                                raise ValueError(
                                    "ubuntu_sso_email is required "
                                    "when initial login using console-conf"
                                )
                            provision_plan["config"][
                                "username"
                            ] = ubuntu_sso_email    
                            break
```
and you'll need to `import contextlib` on top.

( I hope I used the correct indentation, the GitHub editor is a PITA with code )

Reply:
Yes, it can make the code clearer and readable.